core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1,101 +1,381 @@
+
+  def onFinalize[F2[_]](f: F2[Unit])(implicit S: Sub1[F,F2], F2: Monad[F2]): Stream[F2,O] =
+    Stream.bracket(F2.pure(()))(_ => Sub1.substStream(self), _ => f)
+
+  def open: Pull[F, Nothing, Handle[F,O]] = Pull.pure(new Handle(List(), self))
+
+  def output: Pull[F,O,Unit] = Pull.outputs(self)
+
+  def pull[F2[_],O2](using: Handle[F,O] => Pull[F2,O2,Any])(implicit S: Sub1[F,F2]) : Stream[F2,O2] =
+    Sub1.substPull(open).flatMap(h => Sub1.substPull(using(h))).close
+
+  /** Converts a `Stream[Nothing,O]` in to a `Stream[Pure,O]`. */
+  def pure(implicit S: Sub1[F,Pure]): Stream[Pure,O] = covary[Pure]
+
+  /** Repeat this stream an infinite number of times. `s.repeat == s ++ s ++ s ++ ...` */
+  def repeat: Stream[F,O] = self ++ repeat
+
+  def runFree: Free[F,Unit] =
+    runFoldFree(())((_,_) => ())
+
+  def runTraceFree(t: Trace): Free[F,Unit] =
+    runFoldTraceFree(t)(())((_,_) => ())
+
+  def runFoldFree[O2](z: O2)(f: (O2,O) => O2): Free[F,O2] =
     get.runFold(z)(f)
 
-  override final def runFoldTraceFree[O2](t: Trace)(z: O2)(f: (O2,O) => O2): Free[F,O2] =
+  def runFoldTraceFree[O2](t: Trace)(z: O2)(f: (O2,O) => O2): Free[F,O2] =
     get.runFoldTrace(t)(z)(f)
 
-  final def step: Pull[F,Nothing,Step[Chunk[O],Handle[F,O]]] =
+  def runLogFree: Free[F,Vector[O]] =
+    runFoldFree(Vector.empty[O])(_ :+ _)
+
+  /** Alias for `self through [[pipe.prefetch]](f).` */
+  def prefetch[F2[_]](implicit S:Sub1[F,F2], F2:Async[F2]): Stream[F2,O] =
+    Sub1.substStream(self)(S) through pipe.prefetch
+
+  /** Alias for `self through [[pipe.rechunkN]](f).` */
+  def rechunkN(n: Int, allowFewer: Boolean = true): Stream[F,O] = self through pipe.rechunkN(n, allowFewer)
+
+  /** Alias for `self through [[pipe.reduce]](z)(f)`. */
+  def reduce[O2 >: O](f: (O2, O2) => O2): Stream[F,O2] = self through pipe.reduce(f)
+
+  /** Alias for `self through [[pipe.scan]](z)(f)`. */
+  def scan[O2](z: O2)(f: (O2, O) => O2): Stream[F,O2] = self through pipe.scan(z)(f)
+
+  /** Alias for `self through [[pipe.scan1]](f)`. */
+  def scan1[O2 >: O](f: (O2, O2) => O2): Stream[F,O2] = self through pipe.scan1(f)
+
+  def scope: Stream[F,O] =
+    Stream.mk { StreamCore.scope { self.get } }
+
+  /** Alias for `self through [[pipe.shiftRight]]`. */
+  def shiftRight[O2 >: O](head: O2*): Stream[F,O2] = self through pipe.shiftRight(head: _*)
+
+  /** Alias for `self through [[pipe.sliding]]`. */
+  def sliding(n: Int): Stream[F,Vector[O]] = self through pipe.sliding(n)
+
+  /** Alias for `self through [[pipe.split]]`. */
+  def split(f: O => Boolean): Stream[F,Vector[O]] = self through pipe.split(f)
+
+  def step: Pull[F,Nothing,(Chunk[O],Handle[F,O])] =
     Pull.evalScope(get.step).flatMap {
       case None => Pull.done
       case Some(Left(err)) => Pull.fail(err)
-      case Some(Right(s)) => Pull.pure(s.copy(tail = new Handle(List(), Stream.mk(s.tail))))
+      case Some(Right((hd, tl))) => Pull.pure((hd, new Handle(Nil, Stream.mk(tl))))
```

Comment:
I am just thinking if naming the handle a `tl` is correct. i think `h` was pretty ok. I am leaning towards (c,h) where c means chunk, and in case of receive1 simple (o,h). WDYT?